docs/language/ql-training-rst/cpp/bad-overflow-guard.rst

```diff
@@ -0,0 +1,228 @@
+
+.. note::
+
+  - An important part of the query is to determine whether a given expression has a “small” type that is going to trigger integer promotion.
+  - We therefore write a helper predicate for small expressions.
+  - This predicate effectively represents the set of all expressions in the database where the size of the type of the expression is less than 4 bytes, that is less than 32 bits.
+
+QL query: bad overflow guards
+=============================
+
+We can ensure the operands being added have size less than 4 bytes, using our new predicate.
+QL has logical quantifiers like ``exists`` and ``forall``, allowing us to declare variables and enforce a certain condition on them.
+Now our query becomes:
+
+.. rst-class:: build
+
+.. literalinclude:: ../query-examples/cpp/bad-overflow-guard-2.ql
+   :language: ql
+
+.. note::
+
+  - Recall from earlier that what makes an overflow check a “bad” check is that all the arguments to the addition are integers smaller than 32 bits.
+  - We could write this by using our helper predicate ``isSmall`` to specify that each individual operand to the addition ``isSmall`` (that is under 32 bits):
+
+  .. code-block:: ql
+
+    isSmall(a.getLeftOperand()) and
+    isSmall(a.getRightOperand())
+
+  - However, this is a little bit repetitive. What we really want to say is that: all the operands of the addition are small. Fortunately, QL provides a ``forall`` formula that we can use in these circumstances.
+  - A ``forall`` has three parts:
+    - A declaration part, where we can introduce variables.
+    - A “range” part, which allows us to restrict those variables.
+    - A “condition” part. The ``foral`` as a whole holds if the condition holds for each of the values in the range.
+  - In our case:
+    - The declaration introduces a variable for Expressions, called ``op``. At this stage, this variable represents all the expressions in the program.
+    - The “range” part, ``op = a.getAnOperand()``,  restricts ``op`` to being one of the two operands to the addition.
+    - The “condition” part, ``isSmall(op)``, says that the ``forall`` holds only if the condition - that the ``op`` is small - holds for everything in the range - that is both the arguments to the addition
+
+QL query: bad overflow guards
+=============================
+
+Sometimes the result of the addition is cast to a small type of size less than 4 bytes, preventing automatic widening. We don’t want our query to flag these instances.
+
+We can use predicate ``Expr.getExplicitlyConverted()`` to reason about casts that are applied to an expression, adding  this restriction to our query:
+
+.. code-block:: ql
+
+   not isSmall(a.getExplicitlyConverted())
+
+The final query
+===============
+
+.. literalinclude:: ../query-examples/cpp/bad-overflow-guard-3.ql
+   :language: ql
+
+This query finds a single result, which is `a genuine bug in ChakraCore <https://github.com/Microsoft/ChakraCore/commit/2500e1cdc12cb35af73d5c8c9b85656aba6bab4d>`__.
```

Comment:
Possibly worth being more explicit here about the bug being present only in our historic snapshot?

```suggestion
This query finds a single result in our historic snapshot, which was `a genuine bug in ChakraCore <https://github.com/Microsoft/ChakraCore/commit/2500e1cdc12cb35af73d5c8c9b85656aba6bab4d>`__.
```